HW4_Voskoboinikov/ultimate_protein_tools.py

```diff
@@ -0,0 +1,691 @@
+            for rna in rnas:
+                frame = ''
+                for i in range(frame_number, len(rna) - (frame_number + 1), 3):  # set frame-dependent range to iterate
+                    frame_codon = rna[i:i + 3]  # extract codon
+                    if not check_all_upper(frame_codon):  # check if all letters in codon uppercase
+                        frame_codon = frame_codon.lower()  # if not change all to lowercase
+                    frame += RNA_CODON_TABLE[frame_codon]
+                    if is_stop_codon_termination_enabled and RNA_CODON_TABLE[
+                        frame_codon] == '*':  # stop writing if meet stop-codon
+                        break
+                frames_list.append(frame)  # append frame to frames list
+            frameshift_dct[f'frame_{frame_number}'] = list(set(frames_list))  # clean duplicates and write to dict
+        return frameshift_dct
+    return "You don't fucking know what you're doing!"  # politely ask user to reconsider their actions
+
+
+def get_length_of_protein(seq: str, **_) -> int:
+    """
+    Calculates the length of a protein.
+
+    Argument:
+    - seq (str): sequence to calculate the length
+
+    Return:
+    - int: sequence length
+    """
+
+    return len(seq)
+
+
+def count_aa(seq: str, aminoacids: str = None, **_) -> dict:
+    """
+    Counts the number of given or all amino acids in a protein sequence.
+
+    Arguments:
+    - seq (str): sequence to count amino acids
+    - aminoacids (str): which amino acids to count in sequence
+
+    Return:
+    - dict: a dictionary with amino acids and its count
+    """
+
+    aa_dict_count = {}
+    if (aminoacids is None) or (aminoacids == ''):
+        '''
+        I added an additional condition for user-friendly experience.
+        E.g., we can want to find specific aminoacid, look on result and then look on all aminoacids.
+        Without this condition we have to delete keyword argument, but with it we can only make it empty.
+        '''
+        aminoacids = ''.join(set(seq))
+    for aa in aminoacids:
+        aa_dict_count[aa] = seq.count(aa)
+    return aa_dict_count
+
+
+def get_fracture_of_aa(seq: str, show_as_percentage: bool = False, aminoacids: str = None, **_) -> dict:
+    """
+    Calculates the fracture or percentage of amino acids in a protein sequence.
+
+    Arguments:
+    - seq (str): sequence in which you need to calculate the fracture of amino acids
+    - show_as_percentage (bool): change it to True, if you want to get results with percentages
+    - aminoacids (str): the fracture of which amino acids to count in the sequence
+
+    Return:
+    - dict: a dictionary with amino acids and its fracture or percentage
+    """
+
+    if show_as_percentage:
+        mult = 100
+        round_var = 2
+    else:
+        mult = 1
+        round_var = 4
+    aa_dict_count = count_aa(seq, aminoacids=aminoacids)
+    aa_dict_percent = {}
+    len_of_protein = get_length_of_protein(seq)
```

Comment:
```suggestion
    aas_counts = count_aa(seq, aminoacids=aminoacids)
    aa_dict_percent = {} 
    protein_len = get_length_of_protein(seq)
```

`aa_dict_percent` название переменной не соотвествует содержанию, там могут лежать не проценты